### ISSUE TITLE
live/example.com is not updated atomically

### ISSUE THREAD

**Issue by RalphCorderoy** [certbot/certbot#9900]
The updating of the files available under `/live/example.com` is not atomic. Programs can read a mix of incompatible old and new files via the symlinks or find a symlink isn't present. The programs using the files shouldn't need to workaround the lack of atomicity.

One way to remove the burden from them is to leave `/live` as an area only used by Certbot and have a deploy hook on renewal which atomically updates copies outside `/etc/letsencrypt` which the programs use. For example, have a symlink to a directory which the program _open(2)s_ and then uses _openat(2)_ or similar to read the files in that directory rather than have the path repeatedly followed.

On a heavily loaded system the window of inconsistency during a renew can be quite long. It is made worse by the Python code not being structured to minimise its length. If programs are started by a socket connection to handle one transaction, with many domains and many sockets, then the files are being read frequently which raises the chance of hitting the race-condition window.

Ideally, Certbot should make an atomic update available but assuming that won't happen then the issue should be documented so users investigating odd errors may save time.

**Comment by RalphCorderoy** [comment]
I see [Postfix]( https://www.postfix.org/TLS_README.html) is aware of the general problem.

> You can also store the keys separately from their certificates, again provided each is listed before the corresponding certificate chain. Storing a key and its associated certificate chain in separate files is not recommended, because this is prone to race conditions during key rollover, as there is no way to update multiple files atomically.

**Comment by wgreenberg** [comment]
Unfortunately, you're right that there's a chance for a race condition here, but the structure and behavior of `/live` is deeply baked into the core of how Certbot works, so what you're describing is a fairly major rework of Certbot. I don't believe this possibility is significant or common enough to warrant documenting, though -- if that turns out not to be the case, I'd be glad to add documentation.

**Comment by RalphCorderoy** [comment]
I disagree. Postfix have clearly suffered the problem enough to document it, and we experience it more than once despite having only a handful of systems. Perhaps others either don't diagnose the problem, just retry, or add a bodge to work around it, and not bother reporting any of it upstream.

Without shifting from the /live structure, I explained when opening the issue that Certbot's current code increases the chance of the fault and the time window of inconsistency could be lessened.
